src/Operation/Equals.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/**
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace loophp\collection\Operation;
+
+use Closure;
+use Generator;
+use Iterator;
+
+/**
+ * @immutable
+ *
+ * @template TKey
+ * @template T
+ */
+final class Equals extends AbstractOperation
+{
+    /**
+     * @pure
+     *
+     * @return Closure(Iterator<TKey, T>): Closure(Iterator<TKey, T>): Generator<int, bool>
+     */
+    public function __invoke(): Closure
+    {
+        return
+            /**
+             * @param Iterator<TKey, T> $other
+             *
+             * @return Closure(Iterator<TKey, T>): Generator<int, bool>
+             */
+            static function (Iterator $other): Closure {
+                /**
+                 * @param Iterator<TKey, T> $self
+                 *
+                 * @return Generator<int, bool>
+                 */
+                return static fn (Iterator $self): Generator => yield (iterator_count($other) === iterator_count($self))
+                        && yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($self);
```

Comment:
@drupol I realised it's a bit trickier than I thought initially, can you think of how to do this better and still pass all the tests? (don't mind the current failures, it's just that some cannot be run on php 7.4 so I'll need to implement that)

Reply:
It's the same issue as in the `Combine` operation. I'll give it a try later for sure!

Reply:
I got the tests working by doing:

```diff
diff --git a/src/Operation/Equals.php b/src/Operation/Equals.php
index b758352a..4f71f4a4 100644
--- a/src/Operation/Equals.php
+++ b/src/Operation/Equals.php
@@ -40,8 +40,10 @@ final class Equals extends AbstractOperation
                  *
                  * @return Generator<int, bool>
                  */
-                return static fn (Iterator $self): Generator => yield (iterator_count($other) === iterator_count($self))
-                        && yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($self);
+                return static function (Iterator $self) use ($other): Generator {
+                    return yield (iterator_count($other) === iterator_count($self))
+                    && Pipe::of()(Diff::of()(...$other), IsEmpty::of())($self)->current();
+                };
             };
     }
 }
```

Reply:
ooh good point, I didn't know about that one. I'll see if I can refactor a bit based on that

Reply:
why is that one more efficient? We are still doing `iterator_count` on both of them. Let me know what you think of this version:

https://github.com/loophp/collection/pull/152/commits/11611a9170de8cc0d817d5a2ffec6c16736e4273

Reply:
I didn't say that my fix was optimized.

When I need to do something like this, I always proceed by "iteration" (pun intended!), I start with a dumb/unoptimized version, and then I refine it multiple times.

Check how I would see the thing here:

```diff
diff --git a/src/Operation/Equals.php b/src/Operation/Equals.php
index 643719c8..222c43ed 100644
--- a/src/Operation/Equals.php
+++ b/src/Operation/Equals.php
@@ -40,21 +40,19 @@ final class Equals extends AbstractOperation
                  *
                  * @return Generator<int, bool>
                  */
-                return static function (Iterator $self) use ($other): Generator {
+                return static function (Iterator $iterator) use ($other): Generator {
                     $sameCount = true;
 
-                    do {
-                        if ($self->valid() !== $other->valid()) {
-                            $sameCount = false;
-
-                            break;
-                        }
-
-                        $self->next();
+                    while ($other->valid() && $iterator->valid()) {
+                        $iterator->next();
                         $other->next();
-                    } while ($self->valid() || $other->valid());
+                    }
+
+                    if ($other->valid() !== $iterator->valid()) {
+                        $sameCount = false;
+                    }
 
-                    yield $sameCount && yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($self);
+                    return yield $sameCount && yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
                 };
             };
     }
```

I'm unsure about the last line.

Should we do: 

```php
return yield $sameCount && yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
```

or 

```php
return yield $sameCount && Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator)->current();
```

It would be very interesting to check if the second `yield` is triggered when `$sameCount` is `false`.

Reply:
The only thing I don't like in this thing is the `Diff::of()(...$other)`, we really should find a better way, I have the impression that this is absolutely not the best option.

Reply:
> I'm unsure about the last line.

I like the `yield from` version

Reply:
It's not a matter of `yield` or `yield from`, I also prefer the `yield from`.

I'm afraid that those are as we have 2 yields, they are not evaluated as a real `Logical AND`.

Anyway, let me draft a quick example to prove my point.

Reply:
And that's what I though: https://3v4l.org/t7Nqk

Reply:
> Check how I would see the thing here:
Thank you for this suggestion, I simplified it even more and I think it looks good now

Reply:
No, we cannot have 2 yields inside the function.

2 yields equals 2 returned value.

We must have only 1 yield.

Reply:
> The only thing I don't like in this thing is the Diff::of()(...$other), we really should find a better way, I have the impression that this is absolutely not the best option.

I think you might have the right feeling there, but we can always improve on it later.

The reality is that with the way I envision `equals` working, we have to check all elements of both collections. I think it shouldn't be a problem as long as we mention in the documentation that there is a performance cost to it.

When we add `same` we will clarify that this is the more efficient one and the one to use if you want to check for `===` rather than just seeing if both collections contain the same elements.

Reply:
Oh wait. I just noticed how you did it in `Collection` class.

Reply:
Brilliant ! It's smart and beautiful, nice, I love it !

Reply:
The only thing I would change:

```diff
diff --git a/src/Operation/Equals.php b/src/Operation/Equals.php
index d3b2b675..d0bae356 100644
--- a/src/Operation/Equals.php
+++ b/src/Operation/Equals.php
@@ -47,10 +47,10 @@ final class Equals extends AbstractOperation
                     }
 
                     if ($other->valid() !== $iterator->valid()) {
-                        yield false;
+                        return yield false;
                     }
 
-                    yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
+                    return yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
                 };
             };
     }
```

Like that, if the operation is used standalone, it won't yield 2 values, it will only yield 1 value.

Reply:
> No, we cannot have 2 yields inside the function.
> 
> 2 yields equals 2 returned value.
> 
> We must have only 1 yield.

Ah you're right, I missed that 🤦‍♂️ . The operation wouldn't be properly usable by itself

Reply:
No no, you should revert your [last commit](74b0e180f4736a90ecd932d4554310b94e1d3da2) and use [this diff here](https://github.com/loophp/collection/pull/152#discussion_r671887211).

Reply:
> The only thing I would change:
> 
> ```diff
> diff --git a/src/Operation/Equals.php b/src/Operation/Equals.php
> index d3b2b675..d0bae356 100644
> --- a/src/Operation/Equals.php
> +++ b/src/Operation/Equals.php
> @@ -47,10 +47,10 @@ final class Equals extends AbstractOperation
>                      }
>  
>                      if ($other->valid() !== $iterator->valid()) {
> -                        yield false;
> +                        return yield false;
>                      }
>  
> -                    yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
> +                    return yield from Pipe::of()(Diff::of()(...$other), IsEmpty::of())($iterator);
>                  };
>              };
>      }
> ```
> 
> Like that, if the operation is used standalone, it won't (_potentially, when the first one is false_) yield 2 values, it will only yield 1 value.

Sadly you were right before and this also doesn't fix it:
https://3v4l.org/bJaaW#v8.0.8

However, the latest change fixes it. Based on:
https://3v4l.org/VMkTJ#v8.0.8

Reply:
Ah the `return yield` 🎉

Reply:
Yep, early returns are the way to go to avoid creating intermediary variables! https://3v4l.org/9rF6j#v8.0.8

Reply:
excellent, thanks for the assistance on this